### DISCUSSION TITLE
Python API: Issue with lazy evaluation when iterating (huge query plans produced). Need advice.

### DISCUSSION THREAD

**Discussion by lionelperrin** [duckdb/duckdb#14862] [Q&A]
Hello,

We must update multiple times (x1000 times) a large collection. Each time, we update the same columns `state1, state2, ...` of the collection. So that the sql code can be maintained, we'd like to use some intermediate columns to compute the new state.
Here is the equivalent pseudo python code:
```python
collection_n = collection_0 # initial state
for i in range(1000):
  collection_n_plus_1 = duckdb.sql(
  """
  select
    -- many and large constant columns
    id, rate_values, ...
    -- updated result (f1, f2, .. are sql expressions)
    f1(intermediate1, intermediate2, ...) as state1,
    f2(intermediate1, intermediate2, ...) as state2,
    ...
  from (
    select
      -- many and large constant columns
      id, rate_values, ...,
      -- intermediate columns, required to calculate the new state
      xxx as intermediate1, yyy as intermediate2, ....
   from
      collection_n
    )
   """
   )
  collection_n = collection_n_plus_one
 ```
 
 I face very poor performances with such an implementation (approx 1800s in my case where the equivalent code in pure python takes 300s): When iterating over these definitions, the sql query for `collection_1000` has a very complex query plan, causing issues with the query planner.

I've then tried to materialize results at each step.
```
create or replace temp table collection_n as
select -- same select as before
  ...
```
The performances are better (approx 280s) but still not at the expected level. I suspect that it is due to the large constant columns being copied.

An update statement like the following one is better (approx 15s), but because of the need for intermediate columns, it uses a Hash join, which prevents optimal performances.
```
update collection_n
set
  state1 = collection_n_plus_1.state1,
  state2 = collection_n_plus_1.state2
from
   ( ...) collection_n_plus_1 -- same select as before
 where
   collection_n.id =  collection_n_plus_1.id
```

Any thoughts how to solve this ? Is there a way to write the update statement without a join ? Is there a way to materialize tables without copying columns from another table when identical ?

**Comment by Tishj** [comment]
I'm not following, you're talking about lazy evaluation but there is no mention of a consuming method?
`fetchmany()` will make the internal result streaming, which means the query execution gets intermittently paused when it has filled its buffer, which means you can process the results in a streaming fashion instead of having to wait for the entire result set to be materialized.

**Comment by lionelperrin** [comment]
Thanks @Tishj for your reply.

Here is a simplified code that highlights the issue:
```python
ollection_n = duckdb.sql(
    """
    select
      id, --integer
      values, -- double[120]
      0::double as state
    from collection
    """
)

for i in range(args.repeat):
    collection_n1 = duckdb.sql(
        f"""
        select id, values, intermediate1 + intermediate2 as state from (
          select
            id, values,
            -- INTERMEDIATE CALCULATIONS
            -- in the real use case, formulas are more complex and 
            -- the use of intermediate columns helps at keeping thing readable
            values[{i % 120 +1}] as intermediate1,
            values[{(i+1) % 120 +1}] as intermediate2
          from collection_n
        )
        """
    )
    collection_n = collection_n1

collection_n.to_parquet('/tmp/result.parquet')
```
if I execute this, we can see that there is something wrong with the performances which are not linear with the number of iterations:
```
$> time python -m samples.duckdb.test --repeat 100
1.89s
$> time python -m samples.duckdb.test --repeat 200
8.22s
$> time python -m samples.duckdb.test --repeat 300
21.91s
```

I'm unsure how to best implement these iterations. I feel I should periodically materialize the collection. If I create a table in every loop using `create as select`, it incurs a significant cost, likely due to the large column `values` in my collection. Using an update statement offers greater performance, but it doesn't allow me to use intermediate columns for calculating `state`, unless I perform a Hash join using a sub-SQL query.
```python
duckdb.sql(
    """
    create or replace table collection_n as
    select
      id, --integer
      values, -- double[120]
      0::double as state
    from collection
    """
)

for i in range(args.repeat):
    duckdb.sql(
        f"""
        update collection_n
        set state = collection_n1.intermediate1 + collection_n1.intermediate2
        from (
          select
            id,
            -- INTERMEDIATE CALCULATIONS
            -- in the real use case, formulas are more complex and 
            -- the use of intermediate columns helps at keeping thing readable
            values[{i % 120 +1}] as intermediate1,
            values[{(i+1) % 120 +1}] as intermediate2
          from collection_n
        ) collection_n1
        where collection_n.id = collection_n1.id -- I don't see how to avoid this non necessary hash join
        -- which wouldn't be necessary without intermediate columns
        """
    )

duckdb.sql('select * from collection_n').to_parquet('/tmp/result.parquet')
```

**Reply by Tishj** [reply]
You could put a CTE in there to hold your intermediate columns

**Reply by lionelperrin** [reply]
Thanks @Tishj 

> You could put a CTE in there to hold your intermediate columns

Do you mean there is a way to rewrite this query with a CTE that avoids the join? I can't see how. Here is what I have:
```sql
with collection_n1 as (
  select
    id, values,
    -- INTERMEDIATE CALCULATIONS
    -- in the real use case, formulas are more complex and 
    -- the use of intermediate columns helps at keeping thing readable
    values[{i % 120 +1}] as intermediate1,
    values[{(i+1) % 120 +1}] as intermediate2
  from collection_n
)
update collection_n
set state = collection_n1.intermediate1 + collection_n1.intermediate2
from collection_n1
where collection_n1.id = collection_n.id
```

**Reply by Tishj** [reply]
I wouldn't make that claim, I don't really understand what you're doing here, I am just merely commenting on:
> but it doesn't allow me to use intermediate columns for calculating state

Using a CTE in your select would help with this (a subquery also works, but a CTE is perhaps cleaner)
```sql
update <tbl>
set <...>
from (
    with my_cte as (
        <body containing intermediate calculations>
    )
    select <only the columns you care about>
)
```

**Reply by lionelperrin** [reply]
I'm implementing a benchmark of a calculation engine where we have to repeat some calculations and update a state, iteration after iteration.
The equivalent implementation with pandas could be something as simple as
```python
collection_df = duckdb.sql(
    """
    select
      id, --integer
      values, -- double[120]
      0::double as state
    from collection
    """
).df()

values = np.stack(collection_df['values'].values)
tstart = time.perf_counter()
for i in range(args.repeat):
    # INTERMEDIATE CALCULATIONS
    # in the real use case, formulas are more complex and based on previous state
    # the use of intermediate columns helps at keeping thing readable
    intermediate1 = values[:, i % 120]
    intermediate2 = values[:, (i+1) % 120]
    collection_df['state'] = intermediate1 + intermediate2

duckdb.sql("COPY (select id, state from collection_df) TO '/tmp/result.parquet' (FORMAT 'parquet')")
```

For reference, the pandas implementation takes 0.46s.

For now, the measure performances with duckdb are far from pandas.
 - Without `create as select`: (25s) the query plans become huge and the performances are not linear with the number of iterations
 - With `create as select`: (11.5s) the cost of writing the constant columns (id, values) at every loop is significant
 - With `update`: (5.5s) the where clause (hash join) prevents optimal performances

Would you agree with my conclusions ? Am I missing an obvious way of solving that problem purely with duckdb ? (using pandas is not ideal - we lose duckdb disk spilling)

**Comment by soerenwolfers** [comment]
Can't you simply create the string of the expression of your final column in whatever structured fashion you prefer, in pure python, and then pass that single string as a single column to duckdb to execute on the data?

**Reply by soerenwolfers** [reply]
If not, could you clarify at which points in time you need to access what columns, relative to at what point you know the respective intermediate column definitions?

**Reply by duhlin** [reply]
> Can't you simply create the string of the expression of your final column in whatever structured fashion you prefer, in pure python, and then pass that single string as a single column to duckdb to execute on the data?

In practice, we use about 20 intermediate variables. The logic to compute the result is predefined and remains unchanged between iterations. Therefore, in theory, it should be possible to avoid these intermediate variables. I just realized that it's possible to declare a macro. With this, the join is no longer necessary:
```python
duckdb.sql("create macro intermediate1(values, i) as values[i % 120 +1]")
duckdb.sql("create macro intermediate2(values, i) as values[(i+1) % 120 +1]")

for i in range(args.repeat):
    duckdb.sql(
      f"""
      update collection_n_materialized
      set state = intermediate1(values, {i}) + intermediate2(values, {i})
      """
    )

duckdb.sql('select id, state from collection_n').to_parquet(
    '/tmp/result.parquet')
```
This version takes 1.9s. This is the fastest I have so far with duckdb. Still 4x slower than the pandas version. Do you see other possible optimisations ?

**Reply by soerenwolfers** [reply]
I don't understand why you can't do what I suggested and I don't understand why you need to update anything. From the read pattern you show, it seems you're just interested in the values of a single column that has a complicated expression, as function of the original data. I'm suggesting that you just write down that expression once, in pure Python, and then ask duckdb for its values, once. You can even define all the intermediate expressions as separate columns in a single SELECT clause (duckdb does allow recursive column definitions within a single SELECT clause) if you want, and then wrap that SELECT clause with only one final SELECT that projects the single column you're interested in.

**Reply by soerenwolfers** [reply]
For example:

```python
expression = '1'
for i in range(10):
    expression = f'{expression} * {i} - values[{i}]'
duckdb.query(f'SELECT {expression} FROM tbl')
```

or if you want to be able to inspect intermediate values: 

```python
expressions= {'c0': '1'}
for i in range(10):
    prev = f"c{i}"
    expressions[f'c{i+1}'] = f'{prev} * {i} - values[{i}]'
all = ", ".join(f"{value} AS {key}" for (key, value) in expressions.items()])

duckdb.query(f'SELECT c10 FROM (SELECT {all} FROM tbl)') # inspect by dropping the final projection 
```

**Reply by lionelperrin** [reply]
@soerenwolfers Because the number of iterations is unknown and can be large, the final sql queries can become very large. I've observed that the performances are very poor from approx 30 iterations and not linear with the number of iterations.

This is why in this case, it seems mandatory to persist intermediate results at some point. I've observed better performances with `update` than `create table as select`.

The first test I did is very close to your suggestion. Instead of building the query using string concatenation in python, the query is recursively defined using DuckDBPyRelation.
```python
collection_n = duckdb.sql(
    """
    select
      id, --integer
      values, -- double[120]
      0::double as state
    from collection
    """
)

for i in range(args.repeat):
    # recursive definition of the final request: the query is based on collection_n
    collection_n1 = duckdb.sql(
        f"""
        select id, values, intermediate1 + intermediate2 as state from (
          select
            id, values,
            -- INTERMEDIATE CALCULATIONS
            -- in the real use case, formulas are more complex and 
            -- the use of intermediate columns helps at keeping thing readable
            values[{i % 120 +1}] as intermediate1,
            values[{(i+1) % 120 +1}] as intermediate2
          from collection_n
        )
        """
    )
    collection_n = collection_n1

collection_n.to_parquet('/tmp/result.parquet')
```

**Reply by soerenwolfers** [reply]
does that mean you need results after each iteration?

**Reply by soerenwolfers** [reply]
don't get me wrong, I agree that your first attempt shouldn't be as slow, im just trying to find a workaround

**Reply by lionelperrin** [reply]
I don't need intermediate results. Only final results after all iterations are required.
But I observe that the query plan must remain under a given size.

I've noticed that there is a balance to find between writing intermediate results too frequently and a huge query plan. This implementation is faster (~ 1.0 s) but I don't find it very pleasant to read. Are you aware of an equivalent of spark `.cache()` method ?
```python
duckdb.sql(
  """
  create or replace table collection_n_materialized as
  select id, values, 0::double as state from collection
  """
)

collection_n = duckdb.table('collection_n_materialized')
for i in range(args.repeat):
    # every 25 iterations, persist intermediate results in collection_n_materialized
    if i % 25 == 0:
      duckdb.sql(
        """
        update collection_n_materialized
        set state = collection_n.state
        from collection_n
        where collection_n_materialized.id = collection_n.id
        """
      )
      collection_n = duckdb.table('collection_n_materialized')
    collection_n = duckdb.sql(
      f"""
      select id, values, intermediate1 + intermediate2 as state from (
        select
          id, values,
          -- INTERMEDIATE CALCULATIONS
          -- in the real use case, formulas are more complex and 
          -- the use of intermediate columns helps at keeping thing readable
          values[{i % 120 +1}] as intermediate1, -- not the real formula, just a placeholder
          values[{(i+1) % 120 +1}] as intermediate2 -- not the real formula, just a placeholder
        from collection_n
      )
      """
    )

duckdb.sql('select id, state from collection_n').to_parquet('/tmp/result.parquet')
```

**Reply by soerenwolfers** [reply]
>  I observe that the query plan must remain under a given size.

Size as in "a single long column expression" (as in my first snippet)? Or as in, "many short column expressions" (my second snippet)?  Or as in "number of nested subqueries (some of your suggestions)? 

And "must remain" as in you get an error else? Or a super linear parsing time? Or a super linear execution time?


> This implementation is faster (~ 1.0 s) but I don't find it very pleasant to read

Have you tried my suggestions? What's not working for you with them?

**Reply by soerenwolfers** [reply]
Just to be clear, from what I understood your goal is to compute some complicated but structured column expression, let's assume it's 

```
f(f(f(x, x + 1), f(x, x + 1) +1), f(f(x, x + 1), f(x, x + 1) + 1) + 1)
```

There are four ways to go about this that were discussed so far:

1: `SELECT  f(f(f(x, x + 1), f(x, x + 1) + 1), f(f(x, x + 1), f(x, x + 1) + 1) + 1) FROM df` (my first suggestion: Just write down the expression and select it)

2: `SELECT final FROM (SELECT f(x, x + 1) AS y, f(y, y + 1) AS z, f(z, z + 1) AS final FROM df)` (my second suggestion: write down the components in recursive form, put them all in a single `SELECT`, and then nest it once to get the final expression)

3: Repeatedly `UPDATE` the table with `x := f(x, x + 1)`
(one of your suggestions)

4: Nest repeatedly:
 ```
SELECT f(z, z + 1) as final FROM 
(
    SELECT f(y, y + 1) AS z FROM
    (
        SELECT f(x, x + 1) AS y FROM df
    )
)
```
(another one of your suggestions)

Have you tried 1 and 2?
